feat: Add `set` for `Array<T>`

fslongjin opened the pull request:
Add `set()` function for `Array<T>`

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aya-rs/aya/1005)
<!-- Reviewable:end -->

fslongjin commented:
ping

tamird commented:
> ping

Needs tests.

vadorovsky commented:
To be precise - a great place to add tests would be https://github.com/aya-rs/aya/blob/main/test/integration-ebpf/src/map_test.rs

A simple `set` call on any kind of map with some dummy values would be enough :slightly_smiling_face:

fslongjin commented:
Ok, I'll add tests tomorrow

fslongjin commented:
> To be precise - a great place to add tests would be https://github.com/aya-rs/aya/blob/main/test/integration-ebpf/src/map_test.rs
> 
> A simple `set` call on any kind of map with some dummy values would be enough 🙂

I've already added a test and rebased the branch to the latest version. Could you please review it? Thank you!

tamird reviewed (changes requested):
> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1005)* status: 1 of 3 files reviewed, 4 unresolved discussions (waiting on @fslongjin)

___
*[`ebpf/aya-ebpf/src/maps/array.rs` line 8 at r1](https://reviewable.io/reviews/aya-rs/aya/1005#-OCn3AS-4wREYVbMS9qR:-OCn3AS-4wREYVbMS9qS:b-vyk4mr) ([raw file](https://github.com/aya-rs/aya/blob/5458de0760361f26a10695a2102cda8fe1d50b9f/ebpf/aya-ebpf/src/maps/array.rs#L8)):*
> ```XML
> use crate::{
>     bindings::{bpf_map_def, bpf_map_type::BPF_MAP_TYPE_ARRAY},
>     helpers::bpf_map_lookup_elem,
> ```

can you put `bpf_map_update_elem` here?
___
*[`ebpf/aya-ebpf/src/maps/array.rs` line 78 at r1](https://reviewable.io/reviews/aya-rs/aya/1005#-OCn3HJYA75u0xwdxyL3:-OCn3HJYA75u0xwdxyL4:b-9prtd4) ([raw file](https://github.com/aya-rs/aya/blob/5458de0760361f26a10695a2102cda8fe1d50b9f/ebpf/aya-ebpf/src/maps/array.rs#L78)):*
> ```XML
>     /// Sets the value of the element at the given index.
>     pub fn set(&self, index: u32, value: &T, flags: u64) -> Result<(), c_long> {
>         insert(self.def.get(), &index, value, flags)
> ```

`lookup` above is unsafe, why is this one safe? why do we need the `insert` helper?
___
*[`ebpf/aya-ebpf/src/maps/array.rs` line 92 at r1](https://reviewable.io/reviews/aya-rs/aya/1005#-OCn2lYf63Q-KEMPX95d:-OCn2lYf63Q-KEMPX95e:bevsz30) ([raw file](https://github.com/aya-rs/aya/blob/5458de0760361f26a10695a2102cda8fe1d50b9f/ebpf/aya-ebpf/src/maps/array.rs#L92)):*
> ```XML
>         )
>     };
>     (ret == 0).then_some(()).ok_or(ret)
> ```

why are we making an option only to turn it into a result?

```rust
    match unsafe {
        bpf_map_update_elem(
            def as *mut _,
            key as *const _ as *const _,
            value as *const _ as *const _,
            flags,
        )
    } {
        0 => Ok(())
        ret => Err(ret),
    }
```



_Code quote:_
```XML
    let ret = unsafe {
        bpf_map_update_elem(
            def as *mut _,
            key as *const _ as *const _,
            value as *const _ as *const _,
            flags,
        )
    };
    (ret == 0).then_some(()).ok_or(ret)
```
___
*[`test/integration-test/src/tests/smoke.rs` line 74 at r1](https://reviewable.io/reviews/aya-rs/aya/1005#-OCn4MXI31U3HP62NJPe:-OCn4MXI31U3HP62NJPf:bws54dl) ([raw file](https://github.com/aya-rs/aya/blob/5458de0760361f26a10695a2102cda8fe1d50b9f/test/integration-test/src/tests/smoke.rs#L74)):*
> ```XML
> 
> #[test]
> fn map_set() {
> ```

i don't think this is an appropriate place for this test. this file is called "smoke" because it is testing only loading and attaching programs, and what you're doing here is different. please have a look at the other tests that do more sophisticated things.


<!-- Sent from Reviewable.io -->

tamird commented:
@fslongjin are you planning to work on this?

tamird reviewed (commented):
Now that this is basically a 1-liner, I think we can forgo the tests.

> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1005#-:-OKSlsJO1gQMOnDzQ72e:bou4t2h)* status: 0 of 12 files reviewed, all discussions resolved

___
*[`ebpf/aya-ebpf/src/maps/array.rs` line 8 at r1](https://reviewable.io/reviews/aya-rs/aya/1005#-OCn3AS-4wREYVbMS9qR:-OKSeV2v1xk5Tdu4M5KQ:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/5458de0760361f26a10695a2102cda8fe1d50b9f/ebpf/aya-ebpf/src/maps/array.rs#L8)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you put `bpf_map_update_elem` here?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tamird reviewed (approved):
> Reviewed 9 of 12 files at r2, 2 of 3 files at r3, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1005)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @vadorovsky)



<!-- Sent from Reviewable.io -->
